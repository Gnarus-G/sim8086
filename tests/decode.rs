use std::{io::Write, process::Command};

use insta::assert_display_snapshot;

fn test_with(file: &str) {
    let _ = Command::new("nasm")
        .arg(format!("./{}.asm", file))
        .status()
        .unwrap();

    let app_output = assert_cmd::Command::cargo_bin("sim8086")
        .unwrap()
        .arg(format!("./{}", file))
        .output()
        .map(|out| String::from_utf8(out.stdout).unwrap())
        .unwrap();

    let test_file_path = format!("./{}_test.asm", file);

    let mut test_file = std::fs::File::create(&test_file_path).unwrap();

    write!(test_file, "{}", app_output).unwrap();

    let _ = Command::new("nasm").arg(test_file_path).status().unwrap();

    assert_cmd::Command::new("diff")
        .arg(format!("./{}", file))
        .arg(format!("./{}_test", file))
        .assert()
        .success();

    assert_display_snapshot!(app_output);
}

#[test]
fn movs() {
    test_with("many_register_mov");

    test_with("more_movs");

    test_with("challenge_movs");
}

#[test]
fn add_sub_cmp() {
    test_with("add_sub_cmp");
}
