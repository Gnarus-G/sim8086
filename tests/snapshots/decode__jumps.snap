---
source: tests/decode.rs
description: jnz
expression: app_output
---
bits 16

jne label; 2
jne label; -4
jne label; -6
jne label; -4
je label; -2
jl label; -4
jle label; -6
jb label; -8
jbe label; -10
jp label; -12
jo label; -14
js label; -16
jne label; -18
jnl label; -20
jg label; -22
jnb label; -24
ja label; -26
jnp label; -28
jno label; -30
jns label; -32
loop label; -34
loopz label; -36
loopnz label; -38
jcxz label; -40

